src/index.js

```diff
@@ -1,186 +1,523 @@
-  columnCount: 1,
-  flex: 1,
-  flexGrow: 1,
-  flexPositive: 1,
-  flexShrink: 1,
-  flexNegative: 1,
-  flexOrder: 1,
-  gridRow: 1,
-  gridColumn: 1,
-  fontWeight: 1,
-  lineClamp: 1,
-  lineHeight: 1,
-  opacity: 1,
-  order: 1,
-  orphans: 1,
-  tabSize: 1,
-  widows: 1,
-  zIndex: 1,
-  zoom: 1,
-  fillOpacity: 1,
-  stopOpacity: 1,
-  strokeDashoffset: 1,
-  strokeOpacity: 1,
-  strokeWidth: 1
+  let ret = { [`css-${spec.id}`]: '' }
+  Object.defineProperty(ret, 'toString', {
+    enumerable: false,
+    value () {
+      return 'css-' + spec.id
+    }
+  })
+  ruleCache[spec.id] = ret
+  return ret
+}
+
+function isSelector (key) {
+  const possibles = [':', '.', '[', '>', ' ']
+  let found = false
+  const ch = key.charAt(0)
+  for (let i = 0; i < possibles.length; i++) {
+    if (ch === possibles[i]) {
+      found = true
+      break
+    }
+  }
+  return found || key.indexOf('&') >= 0
+}
+
+function joinSelectors (a, b) {
+  let as = a.split(',').map(a => (!(a.indexOf('&') >= 0) ? '&' + a : a))
+  let bs = b.split(',').map(b => (!(b.indexOf('&') >= 0) ? '&' + b : b))
+
+  return bs
+    .reduce((arr, b) => arr.concat(as.map(a => b.replace(/&/g, a))), [])
+    .join(',')
+}
+
+function joinMediaQueries (a, b) {
+  return a ? `@media ${a.substring(6)} and ${b.substring(6)}` : b
+}
+
+function isMediaQuery (key) {
+  return key.indexOf('@media') === 0
+}
+
+function isSupports (key) {
+  return key.indexOf('@supports') === 0
+}
+
+function joinSupports (a, b) {
+  return a ? `@supports ${a.substring(9)} and ${b.substring(9)}` : b
+}
+
+// flatten a nested array
+function flatten (inArr) {
+  let arr = []
+  for (let i = 0; i < inArr.length; i++) {
+    if (Array.isArray(inArr[i])) arr = arr.concat(flatten(inArr[i]))
+    else arr = arr.concat(inArr[i])
+  }
+  return arr
+}
+
+// mutable! modifies dest.
+function build (dest, { selector = '', mq = '', supp = '', src = {} }) {
+  if (!Array.isArray(src)) {
+    src = [src]
+  }
+  src = flatten(src)
+
+  src.forEach(_src => {
+    if (isLikeRule(_src)) {
+      let reg = _getRegistered(_src)
+      if (reg.type !== 'css') {
+        throw new Error('cannot merge this rule')
+      }
+      _src = reg.style
+    }
+    _src = clean(_src)
+    if (_src && _src.composes) {
+      build(dest, { selector, mq, supp, src: _src.composes })
+    }
+    Object.keys(_src || {}).forEach(key => {
+      if (isSelector(key)) {
+        if (key === '::placeholder') {
+          build(dest, {
+            selector: joinSelectors(selector, '::-webkit-input-placeholder'),
+            mq,
+            supp,
+            src: _src[key]
+          })
+          build(dest, {
+            selector: joinSelectors(selector, '::-moz-placeholder'),
+            mq,
+            supp,
+            src: _src[key]
+          })
+          build(dest, {
+            selector: joinSelectors(selector, '::-ms-input-placeholder'),
+            mq,
+            supp,
+            src: _src[key]
+          })
```

Comment:
Should we remove this? I added tests for this and we add the `webkit` and `ms` prefixes in the babel plugin ([not the `moz` one since the usage of firefox versions that require the prefix is under 1%](http://caniuse.com/#feat=css-placeholder)) for tagged template literals but not for objects, we could _probably_ do it for objects if we wanted to.